gen/emit/EmitJulia.cpp

```diff
@@ -548,27 +592,194 @@ struct EmitJulia : public Emitter
     }
 };
 
+struct JlEmitPack : public Emitter
+{
+    ZCMGen& zcm;
+
+    JlEmitPack(ZCMGen& zcm, const string& fname):
+        Emitter(fname), zcm(zcm) {}
+
+    int emitPackage(const string& packName, vector<ZCMStruct*>& packStructs)
+    {
+        // create the package directory, if necessary
+        vector<string> dirs = StringUtil::split(packName, '.');
+        int havePackage = dirs.size() > 0;
+        string pdname;
+        if (havePackage) {
+            pdname = dirs.at(0);
+        } else {
+            pdname = "";
+        }
+        auto& ppath = zcm.gopt->getString("ppath");
+        string packageDirPrefix = ppath + ((ppath.size() > 0) ? "/" : "");
+        string packageDir = packageDirPrefix + pdname + (havePackage ? "/" : "");
+
+        if (packageDir != "") {
+            if (!FileUtil::exists(packageDir)) {
+                FileUtil::mkdirWithParents(packageDir, 0755);
+            }
+            if (!FileUtil::dirExists(packageDir)) {
+                cerr << "Could not create directory " << packageDir << "\n";
+                return -1;
+            }
+        }
+
+        // write the main package module, if any
+
+        // Keep track of which types were already imported into the module
+        unordered_set<string> moduleJlImports;
+
+        // Keep track of which submodules were already defined. Note that this
+        // set is sorted because we always have to define foo.bar before we
+        // can define foo.bar.baz
+        set<string> moduleJlSubmodules;
+        moduleJlSubmodules.insert(packName);
+
+        FILE *moduleJlFp = nullptr;
+        if (havePackage) {
+            // If we are inside a package, then we need to generate a single
+            // Julia file containing the top-level module and any submodules
+
+            // For a type foo.bar.baz.t1, we will put the module file in
+            // foo.jl
+            vector<string> moduleJlFnameParts;
+            moduleJlFnameParts.push_back(packageDirPrefix);
+            moduleJlFnameParts.push_back(pdname + ".jl");
+            string moduleJlFname = StringUtil::join(moduleJlFnameParts, '/');
+            if (moduleJlFp) {
+                fclose(moduleJlFp);
+                moduleJlFp = nullptr;
+            }
+            if (FileUtil::exists(moduleJlFname)) {
+                // If the module exists already, then we need to parse it and
+                // extract any existing sub-modules and types contained in the
+                // module. We'll include those entries in the final generated
+                // module file.
+                moduleJlFp = fopen(moduleJlFname.c_str(), "r");
+                if (!moduleJlFp) {
+                    perror("fopen");
+                    return -1;
+                }
+                // Read chunks of the existing file line by line
+                while(!feof(moduleJlFp)) {
```

Comment:
While I like the idea of having a single module entry point file,                                    
reading through a previously generated one poses some issues:                                        
- if you rerun zcmgen, you will end up importing the same types more                                 
  than once                                                                                          
- if you remove a type from a package, there's nothing to go through                                 
  and delete it from the module file                                                                 
- parallel builds will often kick off many zcmgen tasks that could                                   
  operate on files from the same package, you'll get data races in                                   
  the reading / writing of it in that case.                                                          
I'm wondering if there would be a better way to keep that file accurate,                                
maybe we could add a specific runtime option to the julia section                                    
of zcmgen that updated that file and required the user to pass the                                   
complete set of zcmtypes for any modules represented.

Reply:
> if you rerun zcmgen, you will end up importing the same types more
than once

I'm not sure that's true. I parse the existing module file and store its imports in the ordered set `moduleJlSubmodules` and then update that set as necessary, so there should be no duplicates. 

> if you remove a type from a package, there's nothing to go through
and delete it from the module file

True. I'm not sure how to resolve that. 

> parallel builds will often kick off many zcmgen tasks that could
operate on files from the same package, you'll get data races in
the reading / writing of it in that case.

Agreed that this is a problem. Doesn't this concern (and the above one) also apply to the `__init__.py` file in the Python bindings?

Reply:
Totally possible that Python has this issue too :smiley: ... I'll have to go take a peak. I think the best way around these issues would be to require a single invocation of `zcm-gen` with a specific flag (something like `--julia-module-file` or similar) that requires the complete set of zcmtypes from each represented module to be on the same command line (note that nodejs already requires this). Then we could simply regenerate the entire file without worrying about parallel build nodes or stale packages.

I should have some time in the next couple days to take a stab at that implementation (unless you are raring to go on it :smiley:) so that we can see how we like it.

Reply:
I think that's totally reasonable. In fact, if the entire module contents are known when `zcm-gen` is invoked, then we can probably eliminate all of this `@eval` complexity and just do 

```julia
module PackageName

include("PackageName/type1.jl")
include("PackageName/type2.jl")
```

and then in `type1.jl` do:

```julia
type type1
   ...
end
 ....
```

Reply:
Right, that would require a module file for each nested module, correct? So we'd end up having something like :

Foo.jl
```
module Foo

include("Foo/Bar.jl")
include("Foo/type1.jl")

end
```

Foo/Bar.jl
```
module Bar

include("Foo/Bar/type2.jl")

end
```

Foo/type1.jl
```
type type1
    ...
end
...
```

Foo/Bar/type2.jl
```
type type2
    ...
end
...
```

That seem like an ok structure to you?

Reply:
That looks great to me. The only potential issue is that the order of includes must reflect the dependencies between types.